### DISCUSSION TITLE
Changing the algorithm of `random.h`

### DISCUSSION THREAD

**Discussion by teo-tsirpanis** [dotnet/runtime#65450] [Q&A]
I noticed that the CoreCLR's random number generator in [`random.h`](https://github.com/dotnet/runtime/blob/c1dba415f397853bde5383d07b1c38b0ad072316/src/coreclr/inc/random.h) is using the original partially superseded algorithm of `System.Random`.

While in the BCL we couldn't fully throw the old one away due to compatibility, the runtime's RNG is not a user-facing component but an internal implementation detail. I am interested in trying to replace it with Xoshiro128 or 256, like the BCL when unseeded but _fully_, meaning that same seeds will now return different results, and I am wondering if there any compatibility concerns I should be mindful of.

Perhaps @jkotas knows the answer.

**Comment by danmoseley** [comment]
Is it used seeded in CLR tests? (@trylek)

**Reply by Clockwork-Muse** [reply]
For the case of seeded tests, do we care if the seed changes _just once_ (effectively)?  After the change, the results should be consistent between runs.  If the tests were expecting specific entries to be hit, those should probably have been pulled out as specific data cases.

**Reply by danmoseley** [reply]
I had in mind any existing issues where the repro steps might include a seed using the current algorithm. However I won't speculate further since I know zero about this part of the repo 😄

**Comment by jkotas** [comment]
There are no compat issues to be worried about with this algorithm.

Majority of uses of this random number generator is for various runtime stress modes. Some runtime stress modes have their own local pseudo random number generator (e.g. look for StressRNG in gc.cpp).

If you believe that it is better to have our own implementation of random number generator instead of just depending on `rand()` from C runtime library, it would be nice to de-duplicate the multiple different C/C++ pseudo random number generation facilities in the repo:
- Create shared implementation in C (ie not C++) under src/native
- Replace https://github.com/dotnet/runtime/blob/main/src/coreclr/inc/random.h and other pseudo random generators in the repo with it
- Delete rand and srand from CoreCLR PAL: https://github.com/dotnet/runtime/search?q=PAL_rand
